lib/tilecanvaspaneitem.cpp

```diff
@@ -43,71 +43,40 @@ TileCanvasPaneItem::~TileCanvasPaneItem()
-    const int tilesDown = tilesetProject->tilesHigh();//qMin(mProject->tilesHigh(), qCeil(height() / zoomedTileSize.width()) + 1);
-
-    // Draw the checkered pixmap that acts as an indicator for transparency.
-    // We use the unbounded canvas size here, otherwise the drawn area is too small past a certain zoom level.
-    painter->drawTiledPixmap(0, 0, zoomedTileSize.width() * tilesAcross, zoomedTileSize.height() * tilesDown, mCanvas->mCheckerPixmap);
-
-    for (int y = 0; y < tilesDown; ++y) {
-        for (int x = 0; x < tilesAcross; ++x) {
-            const QPoint topLeftInScene(x * tilesetProject->tileWidth(), y * tilesetProject->tileHeight());
-            const QRect rect(x * zoomedTileSize.width(), y * zoomedTileSize.height(),
-                zoomedTileSize.width(), zoomedTileSize.height());
-
-            // If the tile pen is in use and it's over this tile, draw it, otherwise draw the current tile
-            // at that location as usual.
-            bool previewTile = false;
-            if (tileCanvas->mTilePenPreview) {
-                QRect tileSceneRect(topLeftInScene.x(), topLeftInScene.y(), tilesetProject->tileWidth(), tilesetProject->tileHeight());
-                previewTile = tileSceneRect.contains(tileCanvas->cursorSceneX(), tileCanvas->cursorSceneY());
-            }
-
-            if (previewTile) {
-                painter->drawImage(rect, *tileCanvas->mPenTile->tileset()->image(), tileCanvas->mPenTile->sourceRect());
-            } else {
-                const Tile *tile = tilesetProject->tileAt(topLeftInScene);
-                if (tile) {
-                    painter->drawImage(rect, *tile->tileset()->image(), tile->sourceRect());
-                }
-            }
-
-            if (tileCanvas->mGridVisible) {
-                QPen pen(tileCanvas->mGridColour);
-                painter->setPen(pen);
+    const QSize zoomedTileSize = tilesetProject->tileSize() * mPane->integerZoomLevel();
 
-                painter->drawLine(rect.x(), rect.y(), rect.x(), rect.y() + rect.height() - 1);
+    const QRect tilesRect{QPoint{0, 0}, QSize{tilesetProject->tilesWide(), tilesetProject->tilesHigh()}};
 
-                if (x == tilesAcross - 1) {
-                    // If this is the right-most edge tile, draw a line on the outside of it.
-                    painter->drawLine(rect.x() + zoomedTileSize.width(), rect.y(),
-                        rect.x() + zoomedTileSize.width(), rect.y() + rect.height() - 1);
-                }
-
-                painter->drawLine(rect.x() + 1, rect.y(), rect.x() + rect.width() - 1, rect.y());
-
-                if (y == tilesDown - 1) {
-                    // If this is the bottom-most edge tile, draw a line on the outside of it.
-                    painter->drawLine(rect.x(), rect.y() + zoomedTileSize.height(),
-                        rect.x() + rect.width(), rect.y() + zoomedTileSize.width());
-                }
+    if (tileCanvas->mGridVisible) {
```

Comment:
Can you please briefly explain what changed here?

Reply:
Well the tile preview is now handled by the undo system based  general tool preview, so I removed that and just optimised the rest of the code a little: instead of checking if it's an edge tile each iteration, just draw the remaining edges at the end.

Reply:
Nice!

I did find a weird issue with the tile preview when hovering over the same tile as the one that is selected, will try to write a test case for it.